### PR TITLE
Add EntityPageTrait

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Related to https://github.com/civicrm/civicrm-core/pull/14184

There is lot's of inconsistent but shared code on pages.  This does what we did for forms (with EntityFormTrait) but for pages.  This PR converts `CRM_Contact_Page_View_Relationship` and adds the new trait.

Before
----------------------------------------
No easy way to share common "page" functionality without adding to `CRM_Core_Page`

After
----------------------------------------
Every page that is working with an entity could be converted to use `EntityPageTrait` bringing consistency to parameters / assignments.

Technical Details
----------------------------------------
Basically per EntityFormTrait where appropriate.  Note this allows us to deprecate the dual `$this->_contactId` and `$this->_contactID` that we have on some pages.

Comments
----------------------------------------
@eileenmcnaughton I imagine this is one you will like in some way.

### CODE REVIEW

File: CRM/Contact/Page/View/Relationship.php [hunk 1]
```diff
@@ -217,17 +230,8 @@ public function run() {
...
-
-    if ($context == 'dashboard') {
-      $cid = CRM_Utils_Request::retrieve('cid', 'Integer',
-        $this, FALSE
-      );
-      $url = CRM_Utils_System::url('civicrm/user',
-        "reset=1&id={$cid}"
-      );
+    if ($this->getContext() == 'dashboard') {
+      $url = CRM_Utils_System::url('civicrm/user', "reset=1&id={$this->getContactId()}");
```
Comment (eileenmcnaughton) [inline CRM/Contact/Page/View/Relationship.php]: I don't know if a function call inside curly places in a quote is valid php - I'm guessing maybe it is - but stylistically we have no precedent for it

**Comment by eileenmcnaughton** [comment]
@mattwire I worked through this & it all looks good. I loaded the relationship view form & it seemed unchanged.

I just made one minor comment & with that solved this is mergeable IMHO

File: CRM/Contact/Page/View/Relationship.php (hunk 1 above)
Comment (colemanw) [reply CRM/Contact/Page/View/Relationship.php]: I think it's new to php5, but it works fine; I've done it myself in some spots.

https://www.php.net/manual/en/language.types.string.php#language.types.string.parsing.complex

File: CRM/Core/Page/EntityPageTrait.php [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  }
+
+  /**
+   * Is the form being used in the context of a edit.
+   *
+   * @return bool
+   */
+  protected function isEditContext() {
+    return ($this->getAction() & (CRM_Core_Action::UPDATE | CRM_Core_Action::ADD));
+  }
```
Comment (colemanw) [inline CRM/Core/Page/EntityPageTrait.php]: The `isDeleteContext` et al. are confusing to me because the comments mention a form, but isn't this a page class?

**Comment by colemanw** [comment]
@mattwire I'm curious to know how many page classes you've tried this with to see how broadly applicable it is and how many quirky differences would need resolved.

File: CRM/Contact/Page/View/Relationship.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Contact/Page/View/Relationship.php]: ok - I guess it's ok then - will merge

File: CRM/Core/Page/EntityPageTrait.php (hunk 2 above)
Comment (eileenmcnaughton) [reply CRM/Core/Page/EntityPageTrait.php]: sigh - our separation of concerns huh

**Comment by eileenmcnaughton** [comment]
@colemanw I suspect quite a few - as we've seen with work on the entity form trait & setting form trait - but a lot of the resolutions have been stdisation & metadata cleanup & moving stuff out of the BAO - all the same things we need to do to prepare for a new form layer
